### ISSUE TITLE
Releasing the release character or not?

### ISSUE THREAD

**Issue by henkish** [DonZoeggerle/ediFabric#4]
Hi!

I had an issue where an EDI partner complained that question marks wasn't escaped in our texts.
I have tried to find out if the release character should be escaped or not, this is what unece.org says:

> ?  immediately preceding one of the characters ' + : ?   restores their normal meaning.
> e.g. 10?+10=20 means 10+10=20. Question mark is represented by ??.

So it seems that the EDI partner is correct, when I look in the Edifabric source code it doesn't look like the release character is escaped, this is what the code looks like:

```C#
private bool ContainsTerminator(char value)
{
    return SegmentTerminator.ToCharArray().Contains(value) ||
            ComponentDataElementSeparator.ToCharArray().Contains(value) ||
            DataElementSeparator.ToCharArray().Contains(value) ||
            RepetitionSeparator.ToCharArray().Contains(value);
}   
```

Is there a reason for not escaping the release character?

Regards Henrik

**Comment by DonZoeggerle** [comment]
Hi Henrik,

Could you please attach a sample edi file which fails to escape the release character ? This was a bug originally, but was fixed in release 5.3 (or should have been :)).

btw, ContainsTerminator() doesn't escape anything it just checks if any of the terminators is missing.

Regards,
Kamen

**Comment by henkish** [comment]
Hi!

It's not when parsing an Edifact file. It's when we're generating an ORDERS message and for instance an FTX field contains a questionmark.

So for instance:

```
var ftx = new S_FTX
{
    D_4451_1 = "ZZZ",
    D_4453_2 = "3",
    C_C108 =
    {
        D_4440_1 = "Can you deliver my order next tuesday? Please!"
    },
    C_C107 = null
};
```

The result looks like: ```FTX+ZZZ+3++Can you deliver my order next tuesday? Please!'```
From what I understand it should look like: ```FTX+ZZZ+3++Can you deliver my order next tuesday?? Please!'```

We use version 5.0 (I think) - I tried to do a quick test upgrade to 5.3, but the output looked the same in my test.

Regards Henrik

**Comment by DonZoeggerle** [comment]
Hi Henrik,

This is not supported at the moment. You are free to escape the free text that goes to the data elements. The escape character for Edifact is ? by default unless otherwise specified in the UNA segment.
For your example  - you can write a method like:

string Escape(string source, InterchangeContext context)
        {
            return source.Replace(context.SegmentTerminator, context.ReleaseIndicator + context.SegmentTerminator)
                .Replace(context.ComponentDataElementSeparator, context.ReleaseIndicator + context.ComponentDataElementSeparator)
                .Replace(context.DataElementSeparator, context.ReleaseIndicator + context.DataElementSeparator)
                .Replace(context.RepetitionSeparator, context.ReleaseIndicator + context.RepetitionSeparator)
                .Replace(context.ReleaseIndicator, context.ReleaseIndicator + context.ReleaseIndicator);
        }

and call it like:

var ftx = new S_FTX
{
    D_4451_1 = "ZZZ",
    D_4453_2 = "3",
    C_C108 =
    {
        D_4440_1 = Escape("Can you deliver my order next tuesday? Please!", interchangecontext)
    },
    C_C107 = null
};

Regards,
Kamen

**Comment by DonZoeggerle** [comment]
...it's a pseudo code, you should probably replace the release separator first and then the rest, but that's the rough idea.

**Comment by henkish** [comment]
Hi again!

Other special characters like "+" and ":" already seems to be escaped by your framework automatically when generating a message. And in general I don't like the idea of escaping of texts in the application - I think text escaping belongs in the framwork. 

So in my opinion I still think this is a bug that needs to be fixed.
Method "ContainsTerminator" only seems to be used in method "EscapeLine" - and by adding ReleaseIndicator to list of terminators the problem seems to be solved.

TLDR: EscapeLine method should escape ReleaseIndicator in addition to terminators (according to http://www.unece.org/fileadmin/DAM/trade/untdid/texts/d423.htm (see 8.2.2)

**Comment by DonZoeggerle** [comment]
Thanks for the update, seems like you've done your homework. I didn't recall it already escapes the rest of the separators lot, makes sense to add '?' to the ContainsTerminator() if it resolves the issue. I hope the unit tests still pass, but anyway - good finding/resolving, well done!
I'll update the code sometime this week.

Cheers,
Kamen

**Comment by DonZoeggerle** [comment]
btw how do you find ediFabric - did it help you to save time on custom edi parsing\generation ?

**Comment by henkish** [comment]
Thanks!

I was recommended it from a former colleague when we were investigating how to implement Edifact D96A. We have been using it for both generating and parsing for over a year now, and it has served us well!

Regards Henrik

**Comment by henkish** [comment]
Hi again!

Think I found another issue.
When generating Edifact and the line ends with + (plus sign) it's removed, and results in invalid Edifact. For instance, when writing text: "Hello world+", it's escaped correctly as "Hello world?+". But then, afterwards, there is code in SegmentParser that's removing the trailing DataElementSeparator (line 327):

``` C#
// Remove the trailing composite terminator
return line.TrimEnd(interchangeContext.ComponentDataElementSeparator.ToCharArray());
```

So it builds the string as: "Hello world?++", then removes the trailing plus sign(s), resulting in "Hello world?". And since ? is the release character it results in invalid Edifact.

I think the fix is to remove only the last + sign at the end of the line:

``` C#
return line.Substring(0, line.Length - interchangeContext.ComponentDataElementSeparator.Length);
```

Regards Henrik

**Comment by DonZoeggerle** [comment]
Thanks Henrik,
I will look into this over the weekend and will prepare a new release including the fix. I think there were other issues with trimming all trailing separators and this has to be revisited.

Thank you again for sharing this.

Regards,
Kamen

**Comment by DonZoeggerle** [comment]
Closing this as the fixes were included in 5.4 release...
